packages/events/src/decorators.ts

```diff
@@ -0,0 +1,286 @@
+          }
+        });
+
+        // Found the payload property but it's referenced indirectly by
+        // a Record or Array, which implies there would be more than 1 payload.
+        if (indexerDepth > 0) {
+          diagnostics.push(
+            createDiagnostic({
+              code: "multiple-event-payloads",
+              messageId: "payloadInIndexedModel",
+              format: {
+                dataPath: stringifyPropertyPath(currentPropertyPath.slice(0, -1)),
+              },
+              target,
+            }),
+          );
+          return;
+        }
+
+        // A payload type was previously found, but only one is allowed.
+        if (payloadType) {
+          diagnostics.push(
+            createDiagnostic({
+              code: "multiple-event-payloads",
+              format: {
+                dataPath: pathToPayload,
+                currentPath: stringifyPropertyPath(currentPropertyPath),
+              },
+              target,
+            }),
+          );
+          return;
+        }
+
+        // A payload property was found.
+        payloadType = prop.type;
+        payloadContentType = getContentType(program, prop);
+        pathToPayload = stringifyPropertyPath(currentPropertyPath);
+      },
+      exitModelProperty() {
+        currentPropertyPath.pop();
+      },
+      tuple(tuple) {
+        currentPropertyPath.push(tuple);
+        // Check if any of the tuple values have already been marked as a payload/payload ancestor.
+        // They won't be visited again, so need to create diagnostic here.
+        tuple.values.forEach((value) => {
+          if (typesThatChainToPayload.has(value)) {
+            diagnostics.push(
+              createDiagnostic({
+                code: "multiple-event-payloads",
+                format: {
+                  dataPath: pathToPayload,
+                  currentPath: stringifyPropertyPath(currentPropertyPath),
+                },
+                target,
+              }),
+            );
+            return;
+          }
+        });
+      },
+      exitTuple() {
+        currentPropertyPath.pop();
+      },
+      model(model) {
+        if (model.indexer) {
+          indexerDepth++;
+        }
+        currentPropertyPath.push(model);
+      },
+      exitModel(model) {
+        if (model.indexer) {
+          indexerDepth--;
+        }
+        currentPropertyPath.pop();
+      },
+    },
+    {},
+  );
+
+  const eventDefinition: EventDefinition = {
+    eventType,
+    root: target,
+    isEventEnvelope: !!payloadType,
+    type: target.type,
+    contentType: envelopeContentType,
+    payloadType: payloadType ?? target.type,
+    payloadContentType: payloadType ? payloadContentType : envelopeContentType,
+  };
+
+  return [eventDefinition, diagnostics];
+}
+
+export function getEventDefinitions(
+  program: Program,
+  target: Union,
+): [EventDefinition[], readonly Diagnostic[]] {
+  const diagnostics: Diagnostic[] = [];
```

Comment:
you can use diagnosticCollector so you can just do diags.pipe() a function that return [value, diagnostics]